File: src/main/java/br/com/warnit/service/CidadeService.java
package br.com.warnit.service;

import br.com.warnit.model.domain.Cidade;
import br.com.warnit.repository.CidadeRepository;
import br.com.warnit.service.exceptions.ObjectNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;

/**
 * @author Guilherme Camargo
 * @since 01/02/2019
 * @version 1.0
 * */
@Service
public class CidadeService {
    @Autowired
    private CidadeRepository cidadeRepository;

    public Cidade findByName(String nomeCidade){
        Optional<Cidade> cidade = cidadeRepository.findByNome(nomeCidade);
        return cidade.orElseThrow(() -> new ObjectNotFoundException("Objeto não encontrado. Nome: "+nomeCidade+
                " Tipo: "+ Cidade.class.getSimpleName()));
    }

    public Cidade findOrPopulate(String nomeCidade, Long id){
        try{
            return findByName(nomeCidade);
        }catch (ObjectNotFoundException e){
            e.printStackTrace();
            return new Cidade(id, nomeCidade, null);
        }
    }
}

File: WarnIt-Address/src/main/kotlin/br/com/warnit/address/service/CidadeService.kt
package br.com.warnit.address.service

import br.com.warnit.address.repository.CidadeRepository
import org.springframework.stereotype.Service
import br.com.warnit.address.model.Cidade
import br.com.warnit.address.service.exception.ObjectNotFoundException


@Service
class CidadeService(private val cidadeRepository: CidadeRepository){

    fun findByName(nome: String): Cidade {
        return cidadeRepository.findByNome(nome)
                .orElseThrow {ObjectNotFoundException("""Objeto não encontrado. Nome: $nome
                     Tipo: ${Cidade::class.simpleName}""")}

    }

    fun findOrPopulate(nome: String, id: Long?): Cidade {
        return try{findByName(nome)}catch (e: ObjectNotFoundException){Cidade(id, nome)}
    }
}